### PR TITLE
update the input `weight` of `_convert_weight_to_int4pack` to `[n][k / 2] uint8`

### PR DESCRIPTION
**Pull request by yanbing-j**
This PR is to update the input `weight` of `_convert_weight_to_int4pack` from `[n][k] int32` to `[n][k / 2] uint8`, both for CPU, CUDA and MPS, which can help decouple int4 model checkpoint with different ISAs and different platforms in `gpt-fast`. The advantage is int4 model checkpoint can be shared in different test machines, without re-generating in one certain platform. Meanwhile, the size of input `weight` can be reduced to `1 / 8`.

Before this PR, packed weight stored in CUDA specific layout: `[n/8][k/(InnerKTiles*16)][32][InnerKTiles/2]`, dtype int32, where InnerKTiles = 2, 4, 8. CPU packed weight viewed as the SAME shape but stored in different layout: `[n/64][k][32]`, dtype uint8. Weight is strongly coupled with platforms (CPU/CUDA) and ISAs (AVX512/AVX2/scalar). And users cannot use a generated weight in another different ISA or platform, because when loading weight into devices, the compute format is different.
image

Now, we use common serialized layout (`[n][k/2] uint8`) for different devices or ISAs as input `weight` of `_convert_weight_to_int4pack`, and each back chooses how to interpret as compute layout.
image

### Performance
Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores)
There is no obvious regression of this PR.
image

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Review by lezcano — APPROVED** [review]
@jgong5 reviewed

**Review by mingfeima — APPROVED** [review]
LGTM. We also need to vectorize the weight packing from cpu backend later on.

**Comment by yanbing-j** [comment]
Hi @malfet , I encountered `compiler internal error` in MPS backend. Could you please help us reach to MPS engineer from Apple? I have an workaround at the end, but I'm not sure whether this is a reasonable change, or it is possible to get a more proper approach. Thanks!

The input of `_convert_weight_to_int4pack` changes from `[n][k] int32` to `[n][k/2] uint8` as mentioned in the PR description. For MPS backend, I try to update this accordingly in https://github.com/pytorch/pytorch/pull/129940/commits/aeb05b2c84be6e5c6ac7a439c793614ecf07fbbf and https://github.com/pytorch/pytorch/pull/129940/commits/3b0891e10d6e25a8bb80ef0de3884690d61344b8. However, it will encounter `compiler internal error` when running the unit test. 

> 2024-07-04T13:07:24.0404840Z =================================== FAILURES ===================================
> 2024-07-04T13:07:24.0406450Z _____ TestLinalgMPSMPS.test__int4_mm_m_1_n_48_q_group_32_num_groups_1_mps ______
> 2024-07-04T13:07:24.0406950Z Traceback (most recent call last):
> 2024-07-04T13:07:24.0407540Z   File "/Users/ec2-user/runner/_work/pytorch/pytorch/test/test_mps.py", line 9191, in test__int4_mm
> 2024-07-04T13:07:24.0408130Z     b_int4pack, b_scales_and_zeros_f32 = convert_weight_to_int4pack(b_f32)
> 2024-07-04T13:07:24.0418080Z   File "/Users/ec2-user/runner/_work/pytorch/pytorch/test/test_mps.py", line 9180, in convert_weight_to_int4pack
> 2024-07-04T13:07:24.0418690Z     b_int4pack = torch._convert_weight_to_int4pack(
> 2024-07-04T13:07:24.0419620Z RuntimeError: Failed to created pipeline state object, error: Error Domain=AGXMetal13_3 Code=3 "Compiler encountered an internal error" UserInfo={NSLocalizedDescription=Compiler encountered an internal error}
> 2024-07-04T13:07:24.0420390Z 
> 2024-07-04T13:07:24.0420570Z To execute this test, run the following from the base repo dir:
> 2024-07-04T13:07:24.0421220Z      python test/test_mps.py -k TestLinalgMPSMPS.test__int4_mm_m_1_n_48_q_group_32_num_groups_1_mps

After adding log print in `_convert_weight_to_int4pack_mps`, the error happens in `id<MTLComputePipelineState> quantizedPSO = lib.getPipelineStateForFunc(kernel);`, after changing input data type of `weight_to_int4pack` in  https://github.com/pytorch/pytorch/pull/129940/commits/3b0891e10d6e25a8bb80ef0de3884690d61344b8#diff-cd1051cc1c1eb45fd317a80160c14c1bf9ce52e4a147301376b95ce42157586eR58.

I made an workaround, trying to change back the input to int32 data type, and update the corresponding code in https://github.com/pytorch/pytorch/pull/129940/commits/4848191670c6deb4e078d1a681848a1560164cd1. The compiler internal error is gone.

I am wondering that, (1) why `uchar` input will lead to internal error? (2) Is there any better approach to fix this error? Thanks!

cc @mingfeima @jgong5

**Comment by yanbing-j** [comment]
@malfet could you please help on https://github.com/pytorch/pytorch/pull/129940#issuecomment-2210467119? Thanks!

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -3255,7 +3255,7 @@ def meta__int_mm(a, b):
 def meta__convert_weight_to_int4pack(w, inner_k_tiles):
     torch._check(w.dim() == 2, lambda: "w must be a 2D tensor")
     torch._check(
-        w.dtype is torch.int32,
+        w.dtype is torch.uint8,
         lambda: f"expected w to be int32, got {w.dtype}",
```
Comment (gau-nernst) [inline torch/_meta_registrations.py:3259]: The error message should be updated to uint8 too.

Comment (yanbing-j) [reply torch/_meta_registrations.py:3259]: Oh, my mistake, thanks for the notification! https://github.com/pytorch/pytorch/pull/130707

**Comment by jerryzh168** [comment]
why non of torchao folds are tagged in the PR? cc @msaroufim @HDCharles

**Comment by jerryzh168** [comment]
Hi @yanbing-j, could you help fix https://github.com/pytorch/ao/pull/504 as well, not sure how we should prepare the weights now

**Comment by yanbing-j** [comment]
Hi @jerryzh168 , I make some fix and ping you on slack. Need your help since I'm not familar with torchao.

**Comment by yanbing-j** [comment]
@jerryzh168

In this PR, what I do is:
1. convert weight `[n][k] int32` to `[n][k/2] uint8` in `_group_quantize_tensor`.
2. change `_convert_weight_to_int4pack` API from `[n][k] int32` to `[n][k / 2] uint8` . The implementation in kernel in each backend will not influence python frontend.

When quantizing the weight, `[n][k/2] uint8` weight needs to be stored, and not related to any platforms or ISAs. When running model, load the `[n][k/2] uint8` weight first, then do `_convert_weight_to_int4pack`, finally run the model.

**Comment by malfet** [comment]
Why this change was not marked as BC breaking? It will require all current users of the API to adjust, wouldn't it?

**Comment by yanbing-j** [comment]
I have no idea about `BC breaking` label. It indeed requires API change in user side.

**Comment by msaroufim** [comment]
Can we revert this PR until the forward fix is made? This PR has been open for a week https://github.com/pytorch/ao/pull/517

@malfet @yanbing-j

**Comment by yanbing-j** [comment]
@msaroufim We update https://github.com/pytorch/pytorch/pull/130915 not to use OpInfo, hope it can help land https://github.com/pytorch/ao/pull/517 ASAP.

**Pull request closed without merging**